python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/ecs.py

```diff
@@ -0,0 +1,250 @@
+
+                # set environment variables for existing overrides
+
+                for container_override in overrides["containerOverrides"]:
+                    container_override["environment"] = container_override.get("environment", [])
+                    container_override["environment"].extend(environment_overrides)
+
+                # set environment variables for containers that are not in the overrides
+                for container_name in all_container_names - container_names_with_overrides:
+                    overrides["containerOverrides"].append(
+                        {
+                            "name": container_name,
+                            "environment": environment_overrides,
+                        }
+                    )
+
+            params: Dict[str, Any] = {
+                "taskDefinition": task_definition,
+                "count": count,
+                "launchType": launch_type,
+                "cluster": cluster,
+                "group": group,
+                "networkConfiguration": network_configuration,
+                "overrides": overrides,
+                "placementConstraints": placement_constraints,
+                "placementStrategy": placement_strategy,
+                "platformVersion": platform_version,
+                "referenceId": reference_id,
+                "startedBy": started_by,
+                "tags": tags,
+                "clientToken": client_token,
+                "volumeConfigurations": volume_configurations,
+                "capacityProviderStrategy": capacity_provider_strategy,
+                "enableECSManagedTags": enable_ecs_managed_tags,
+                "enableExecuteCommand": enable_execute_command,
+                "propagateTags": propagate_tags,
+            }
+
+            params = {
+                k: v for k, v in params.items() if v is not None
+            }  # boto3 doesn't use None as default values
+
+            response = self._client.run_task(**params)
+
+            tasks: List[str] = [task["taskArn"] for task in response["tasks"]]
+
+            try:
+                response = self._wait_for_tasks_completion(tasks=tasks, cluster=cluster)
+
+                # collect logs from all containers
+                for task in response["tasks"]:
+                    task_id = task["taskArn"].split("/")[-1]
+
+                    for container in task["containers"]:
+                        if log_config := log_configurations.get(container["name"]):
+                            if log_config["logDriver"] == "awslogs":
+                                log_group = log_config["options"]["awslogs-group"]
+
+                                # stream name is combined from: prefix, container name, task id
+                                log_stream = f"{log_config['options']['awslogs-stream-prefix']}/{container['name']}/{task_id}"
+
+                                if isinstance(self._message_reader, PipesCloudWatchMessageReader):
```

Comment:
This predates this diff, but why isn't this logic in `read_messages`? Why the instance of check?

Reply:
Because parameters like `log_stream` and `start_time` are only available after the external process has been launched. I'm not sure how can we pass them to `read_messages`. Is there a pattern for something like this available?